src/Components/Components/ref/Microsoft.AspNetCore.Components.netstandard2.0.cs

```diff
@@ -306,6 +333,16 @@ public static partial class HttpClientJsonExtensions
         [System.Diagnostics.DebuggerStepThroughAttribute]
         public static System.Threading.Tasks.Task<T> SendJsonAsync<T>(this System.Net.Http.HttpClient httpClient, System.Net.Http.HttpMethod method, string requestUri, object content) { throw null; }
     }
+    public partial interface IAuthenticationState
+    {
+        bool IsPending { get; }
+        System.Security.Claims.ClaimsPrincipal User { get; }
```

Comment:
Why not represent these two things as `Task<ClaimsPrincipal>`. I get why you are doing this. (In fact our react template does it this way) but in our stack we already respect the notion of Task as a way to represent a "future" value.

If this just becomes a Task, then I can simply await it and react appropriately when it changes.

I think what motivated you to go this way is to provide an idiomatic implementation in AuthorizeView, but I would have used the provider directly on that one (where I want a great degree of control) and used the cascading value in other views (like a menu) where I only care about displaying a current value (like the user name).

In that case, AuthorizeView.razor would be something like
```
if(_currentView != null)
{
    @_currentView
}

Task OnInitAsync()
{
    _currentView = Authorizing;
    var user = await _currentAuthenticationState;
    if(user.IsAuthenticated)
    {
        _currentView = Authorized;
    }
   else
   {
       _currentView = NotAuthorized;
   }
}
```

Reply:
> Why not represent these two things as Task<ClaimsPrincipal>
> If this just becomes a Task, then I can simply await it and react appropriately when it changes.

I don't think it would really simplify the developer's code, because the state is more akin to an observable than a task. The value can change any number of times, at any time, rather than being something you just wait for once then get the final result for all time.

So even if we used `Task`, the framework has to notify you through some non-`Task` based API each time it changes again. Given that we're doing that, we might as well supply the information in a synchronous form. This simplifies developers' code (eliminates asynchrony, which in the default server-side case would always be a no-op anyway), and is especially useful if you're using the state info during rendering independently of `AuthorizeView`.

The alternative form you've posted for `AuthorizeView.razor` is essentially the same as the form in the PR, except substituting `if (_currentAuthenticationState.IsPending)` for `await _currentAuthenticationState`.

Reply:
I've been thinking about this more, and have reached a view that we *should* make use of `Task`, but a bit differently than discussed above. This will not affect usage of `<AuthorizeView>`; it will only affect consumption of the cascading value from `<AuthenticationStateProvider>`.

Reasoning:

Previously, the goal was to make consumption of the cascading value convenient. However, I think there are basically two usage scenarios:

 1. Using the auth state in **procedural logic**
    * For example, in `OnParametersSetAsync`, fetching some data and including some token in the outbound request, or fetching different data depending on who's logged in. The action should be re-triggered whenever the auth state changes.
    * Or for example, in a button click handler, invoking some business-domain operation that depends on the auth state info.
 2. Using the auth state in **rendering output**
    * E.g., `<div>Hello, @AuthState.User.Identity.Name!</div>`. This needs to update if the user logs in or out.
    * Normally you'd use `<AuthorizeView>` instead for this, but there's nothing to stop you from getting the cascaded value directly and using it

Supplying a `Task` of some kind is fine in case 1, because you can trivially `await`. It's even advantageous, because it forces developers to understand that sometimes the info acquisition is async, and they shouldn't just throw "not authorized" errors instantly if they see the user isn't authenticated, because maybe the info is still pending. Another advantage of `Task` here is that it gives some way for "getting auth state failed" errors to surface (i.e., if the `Task` resolves with an exception).

Supplying a `Task` of some kind is disadvantageous in case 2, because it complicates rendering logic. However it doesn't complicate it very much: we already have excellent support for using `await` during lifecycle methods, and people who do async data access already understand how to show "loading..." states while those tasks are in flight.

Since case 2 is unusual anyway (you'd normally use `<AuthorizeView>` for this), it makes sense to give more weight to case 1, which gives a preference for some `Task` instead of an `IsPending` flag.

The other thing that bothered me about the earlier proposal is that it if `IAuthenticationState` contained a `Task<ClaimsPrincipal>`, this would also complicate things for `<AuthorizeView>`. The `context` is of type `IAuthenticationState`, so you'd be in a real mess if you were trying to access something from the claims principal during rendering. However, this isn't really where we'd want to put the task. It's not only the `ClaimsPrincipal` that can change, but rather the *entire* `IAuthenticationState`, which includes any future properties we attach there (we might put antiforgery tokens there too).

So I think we should drop the `IsPending` flag from `IAuthenticationState` but otherwise leave it alone. And then we modify `<AuthenticationStateProvider>` so that the type of the cascading value changes from `IAuthenticationState` to `Task<IAuthenticationState>`. This means that anyone consuming the cascading value is forced to recognise that, sometimes, the state arrives asynchronously. But it doesn't change anything for people using `<AuthorizeView>`, as this already resolves the succession of tasks into a succession of render calls anyway.

Reply:
I think I agree with what you mention here for the most part. I'm not sure there is an scenario for antiforgery tokens here as:
* For server-side rendering the actions/forms/etc would go through the websocket.
* For client-side rendering it is not recommended to use cookies for auth but JWTs which don't require you to use antiforgery tokens.
* In case you need antiforgery tokens:
  * They could surface as claims on the principal.
  * They could be offered by an antiforgery token service, and that way we avoid polluting a type for an uncommon use case.

For these two additional use cases, (access tokens and antiforgery tokens) it would be interesting to see if we could make it trivial for HttpClient to retrieve the required pieces of data and to attach them to outgoing requests automatically for you.

Reply:
> They could surface as claims on the principal

That was my initial design too, but Barry said they should be separate. However you are probably right that we might never even need to expose them in this kind of way.